crates/jstz_core/src/js_fn.rs

```diff
@@ -0,0 +1,136 @@
+use std::{marker::PhantomData, ops::Deref};
+
+use boa_engine::{
+    object::builtins::JsFunction, value::TryFromJs, Context, JsResult, JsValue,
+};
+use boa_gc::{custom_trace, Finalize, Trace};
+
+use crate::value::IntoJs;
+
+pub trait IntoJsArgs<const N: usize> {
+    fn into_js_args(self, context: &mut Context<'_>) -> [JsValue; N];
+}
+
+impl IntoJsArgs<0> for () {
+    fn into_js_args(self, _context: &mut Context<'_>) -> [JsValue; 0] {
+        []
+    }
+}
+
+impl<T0: IntoJs> IntoJsArgs<1> for (T0,) {
+    fn into_js_args(self, context: &mut Context<'_>) -> [JsValue; 1] {
+        [self.0.into_js(context)]
+    }
+}
+
+impl<T0: IntoJs, T1: IntoJs> IntoJsArgs<2> for (T0, T1) {
+    fn into_js_args(self, context: &mut Context<'_>) -> [JsValue; 2] {
+        [self.0.into_js(context), self.1.into_js(context)]
+    }
+}
+
+impl<T0: IntoJs, T1: IntoJs, T2: IntoJs> IntoJsArgs<3> for (T0, T1, T2) {
+    fn into_js_args(self, context: &mut Context<'_>) -> [JsValue; 3] {
+        [
+            self.0.into_js(context),
+            self.1.into_js(context),
+            self.2.into_js(context),
+        ]
+    }
+}
+
+/// A `JsFn<T, N, I, O>` is a `JsFunction` tagged with some Rust types used to handle the `TryFromJs` and `IntoJs` conversions automatically:
+/// - `T` is the type of the `this` parameter;
+/// - `N` is the arity;
+/// - `I` is a tuple `(I1, ..., IN)` that contains the types of the parameters;
+/// - `O` is the type of the output.
+#[derive(Debug)]
+pub struct JsFn<T: IntoJs, const N: usize, I: IntoJsArgs<N>, O: TryFromJs> {
```

Comment:
On reflection, I think the const-level generic `N` adds more effort than it is worth. 

Removing it would give us `JsFn<T, I, O>` which:
- I think is clearer
- It is also the type I'd expect once associated const-types are permitted in Rust (https://github.com/rust-lang/rust/issues/60551), resulting a more localized refactor. 

The downside of such a change is that we'd have return a `Vec` for `IntoJsArgs`. But I think is an acceptable given that it will eventually be changed once associated const-types are in Rust.

Reply:
Also another suggestion would be to have `T` default to `JsValue`: https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#default-generic-type-parameters-and-operator-overloading

Reply:
Hmmm, yeah I guess that is a bit odd. Perhaps we can just wrap over this type with a macro that will substitute `JsValue` for `T` if 2 arguments are given